ibctest gh actions workflow, build and use docker image

agouin opened the pull request:
Adds submodule `ibctest` test case `TestRelayerDocker` to use `NewBuiltinRelayerFactory` with a `CustomDockerImage` built from the latest code.

caching optimizations to the dockerfiles so that
- build deps are cached
- go mods are cached - without go mod changes, will skip straight to build

For the docker image build step, will stream output so that the stdout looks the same as if executing `docker build`. Checks the last line of the docker build output to check if an error occurred. I tested this by adding a compile error and running `make ibctest`, which correctly errors the test and shows the compile error. Upon resolution of the compile error, `make ibctest` ran the docker image build which went straight to the relayer build without having to re-download any deps or go modules, then quickly proceeded into the conformance test.

Adds a separate GH actions workflow to run `make ibctest-docker`

agouin committed:
Build relayer docker image for ibctest

agouin committed:
Add ibctest workflow

agouin committed:
build dep and go mod caching in dockerfile. better docker build error checking

agouin committed:
Use latest ibctest

agouin committed:
Merge branch 'main' into andrew/ibctest_container

mark-rushakoff reviewed (changes requested):
I don't mind using the Docker image as an additive change, but I am not willing to give up the in-process relayer.

I would suggest `make ibctest` retains the current behavior and `make ibctest-docker` uses this new workflow.

agouin committed:
Bring back RelayerFactory local binary impl. add required interface implementations for latest ibctest.

agouin committed:
Merge branch 'main' into andrew/ibctest_container

agouin commented:
> I am not willing to give up the in-process relayer.

Sounds good. I brought back the `relayer` and `relayerFactory` as `Relayer` and `RelayerFactory` in the `ibctest` package (moved from `ibctest_test`), implemented the new/changed interface methods from the latest `github.com/strangelove-ventures/ibctest`:
- `LinkPath` now has opts
- New methods `FlushAcknowledgements`, `FlushPackets`, and `GetWallet`

Added `ibctest-docker` Makefile command, uses that one in gh actions workflow.

mark-rushakoff commented on ibctest/relayer.go:
```diff
@@ -0,0 +1,256 @@
+
+func (r *Relayer) ClearQueue(ctx context.Context, _ ibc.RelayerExecReporter, pathName string, channelID string) error {
+	panic("not yet implemented")
+}
```
`ClearQueue` is no longer part of the ibctest Relayer interface, so this method should be able to be deleted.

agouin replied:
removed

mark-rushakoff commented on ibctest/ibc_test.go:
```diff
@@ -1,296 +1,62 @@
-	return nil
+// TestRelayerBinary runs the ibctest conformance tests against
+// the current state of this relayer implementation built as a binary.
+func TestRelayerBinary(t *testing.T) {
```
This is inaccurate -- there is no standalone binary built in this approach. `TestRelayerInProcess` would be more precise.

agouin replied:
fixed

mark-rushakoff commented on ibctest/relayer.go:
```diff
@@ -0,0 +1,256 @@
+	"go.uber.org/zap/zaptest"
+)
+
+type Relayer struct {
```
This is good, I wanted to move this struct to its own file but never got around to it.

I assume everything here was a direct copy and paste.

agouin replied:
yes, all a copy paste except the new interface methods `FlushAcknowledgements`, `FlushPackets`, and `GetWallet` and `opts` added to `LinkPath`

mark-rushakoff commented on ibctest/relayer_factory.go:
```diff
@@ -0,0 +1,49 @@
+// Build returns a relayer interface
+func (RelayerFactory) Build(
+	t *testing.T,
+	pool *dockertest.Pool,
```
The most recent commit on ibctest (https://github.com/strangelove-ventures/ibctest/commit/9eab603b529ff8ce2a230985f5ae7db05cfab1a6) switched away from dockertest. I think if you pull that in, this just changes to `_ *client.Client` (because the `Relayer` type actually doesn't even use the value).

agouin replied:
updated

mark-rushakoff commented on ibctest/relayer.go:
```diff
@@ -0,0 +1,256 @@
+type Relayer struct {
+	t *testing.T
+
+	pool *dockertest.Pool
```
`pool` is now unused and can be removed.

agouin replied:
removed

mark-rushakoff commented on Makefile:
```diff
@@ -67,7 +67,10 @@ test-short:
 
 ibctest:
-	cd ibctest && go test -race -v .
+	cd ibctest && go test -race -v -run TestRelayer .
```
Looks like this would match both of the test functions -- prefer to match the whole name, taking care to match the updated name.

agouin replied:
updated

mark-rushakoff commented on ibctest/docker.go:
```diff
@@ -0,0 +1,74 @@
+			fmt.Print(logLine.Stream)
+		}
+		if logLine.Aux != nil {
+			fmt.Println(logLine.Aux)
```
These are probably more appropriate as `t.Log` calls.

agouin replied:
done

mark-rushakoff commented on ibctest/docker.go:
```diff
@@ -0,0 +1,74 @@
+
+const RelayerImageName = "ibctestrelayer"
+
+type dockerLogLine struct {
+	Stream      string            `json:"stream"`
+	Aux         any               `json:"aux"`
+	Error       string            `json:"error"`
+	ErrorDetail dockerErrorDetail `json:"errorDetail"`
+}
+
+type dockerErrorDetail struct {
+	Message string `json:"message"`
+}
```
Looks like these overlap with github.com/docker/docker/pkg/streamformatter and github.com/docker/docker/pkg/jsonmessage. It might make more sense to use those, but it's probably fine as-is too.

mark-rushakoff reviewed (approved):
I left some suggestions that can all happen now or later.

agouin committed:
handle feedback
